core/src/main/java/io/questdb/griffin/engine/functions/window/LeadDoubleFunctionFactory.java

```diff
@@ -68,79 +66,51 @@ public Function newInstance(
-            defaultValue = args.getQuick(2);
-            if (!(defaultValue instanceof DoubleFunction)) {
-                throw SqlException.$(argPositions.getQuick(2), "default value must be a double");
+        @Override
+        protected boolean doPass1(Record record, long recordOffset, WindowSPI spi) {
+            double leadValue;
+            if (count < offset) {
+                leadValue = defaultValue == null ? Double.NaN : defaultValue.getDouble(record);
+            } else {
+                leadValue = buffer.getDouble((long) loIdx * Double.BYTES);
             }
-
-            if (defaultValue instanceof WindowFunction) {
-                throw SqlException.$(argPositions.getQuick(2), "default value can not be a window function");
+            double d = arg.getDouble(record);
+            boolean respectNull = !ignoreNulls || Numbers.isFinite(d);
+            if (respectNull) {
+                buffer.putDouble((long) loIdx * Double.BYTES, d);
             }
+            Unsafe.getUnsafe().putDouble(spi.getAddress(recordOffset, columnIndex), leadValue);
+            return respectNull;
         }
-
-        if (offset == 0) {
-            return new LagDoubleFunctionFactory.LeadLagValueCurrentRow(windowContext.getPartitionByRecord(), args.get(0), NAME, windowContext.isIgnoreNulls());
-        }
-
-        if (windowContext.getPartitionByRecord() != null) {
-            Map map = MapFactory.createUnorderedMap(
-                    configuration,
-                    windowContext.getPartitionByKeyTypes(),
-                    LagDoubleFunctionFactory.LAG_COLUMN_TYPES
-            );
-            MemoryARW mem = Vm.getCARWInstance(configuration.getSqlWindowStorePageSize(),
-                    configuration.getSqlWindowStoreMaxPages(), MemoryTag.NATIVE_CIRCULAR_BUFFER
-            );
-
-            return new LeadOverPartitionFunction(
-                    map,
-                    windowContext.getPartitionByRecord(),
-                    windowContext.getPartitionBySink(),
-                    mem,
-                    args.get(0),
-                    windowContext.isIgnoreNulls(),
-                    defaultValue,
-                    offset
-            );
-        }
-
-        MemoryARW mem = Vm.getCARWInstance(
-                configuration.getSqlWindowStorePageSize(),
-                configuration.getSqlWindowStoreMaxPages(),
-                MemoryTag.NATIVE_CIRCULAR_BUFFER
-        );
-        return new LeadFunction(args.get(0), defaultValue, offset, mem, windowContext.isIgnoreNulls());
     }
 
-    static class LeadOverPartitionFunction extends BasePartitionedDoubleWindowFunction {
-        private final Function defaultValue;
-        private final long offset;
-        private final MemoryARW memory;
-        private final boolean ignoreNulls;
-
+    static class LeadOverPartitionFunction extends LeadLagWindowFunctionFactoryHelper.BaseLeadOverPartitionFunction implements WindowDoubleFunction {
```

Comment:
there are warnings in this class (unused value)

Reply:
There's a duplicate code warning, but it's justifiable, there are seems no unused variable warnings.